doc/type_requirements/Hasher.qbk

```diff
@@ -0,0 +1,75 @@
+[/
+    Copyright (c) 2013-2016 Vinnie Falco (vinnie dot falco at gmail dot com)
+
+    Distributed under the Boost Software License, Version 1.0. (See accompanying
+    file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+]
+
+[section:Hasher Hasher]
+
+A type [@Hasher] implements a hash algorithm. This is used to compute the small
+digests NuDB needs to effectively implement a hash table.
+
+The NuDB database stores and retrieves key/value pairs.  This storage is
+organized as "buckets" where each bucket forms the start of a linked list.
+Given a key, one must caluculate which bucket it the key should be
+added to. The library includes code to guarantee that any set of
+keys will map to a more or less even distribution of buckets, thus
+ensuring that the array of buckets is utilized efficiently.  A key
+element in this code is a call to the Hasher function. This is why
+the Hasher function sould be designed to generate a uniform
+distribution of integers for any random collection of keys.
+
+[heading Notation]
+* `H` is a hasher class
+* `h` is an instance of type `H`
+* `s` is a random seed of type `std::uint64_t`
+* `p` denotes a value of type `void const*`
+* `n` denotes a value of type `std::size_t`
+
+[heading Valid Expressions]
+[table
+[[operation] [type] [semantics, pre/post-conditions]]
+[
+    [`H{s}`]
+    [`H`]
+    [
+        `H` is constructed with a seed value integer `s`. To achieve resistance
+        from algorithmic complexity attacks, an implementation of [*Hasher]
+        should ensure that values returned from the hash function will be
+        distinctly different for different values of `s`.
+    ]
+]
+[
+    [`h(p,n)`]
+    [`std::uint64_t`]
+    [
+        Returns the digest of the memory `n` bytes in size and pointed
+        to by `p`. `n` will never be zero. A good hash function will
+        return values with these qualities:
+
+* Values are uniformly distributed in the full range
+
+* Values for the same input are distinctly different for different seeds
+
+* Small changes in the input produce unpredictable output values
+
+    ]
+]
+[
+    [`is_Hasher<H>`]
+    [`std::true_type`]
```

Comment:
You've specified that if anyone creates their own Hasher type, the following statement must return std::true_type.  That's what a trait is. I don't know that I called it a trait but to me that was clearly the intent. (I don't remember if you used the word "trait" or not.

Presumably you created is_Hasher so that your library code can in fact verify that the author of the new implemenation of Hasher is intended to be used in this context and you verify that in the code with something like

static_assert(is_Hasher<H>, "better be a legal model of Hasher concept");

If your not doing this or this is the only thing you're doing with this is_Hasher, I would just drop the trait as it's not really adding anything.
>